custom/icds/repeaters/generators/phi.py

```diff
@@ -26,28 +26,45 @@ def _gender(gender):
         return ""
 
     def get_payload(self, repeat_record, case):
-        data = {
-            "beneficaryname": case.name,
-            "fathername": case.get_case_property('fathers_name') or "",
-            "husbandname": case.get_case_property('husbands_name') or "",
-            "mothername": case.get_case_property('mothers_name') or "",
+        data = self._setup_names(case)
+        data.update({
             "gender": self._gender(case.get_case_property('gender')),
             "villagecode": 442639,
             "subdistrictcode": 3318,
             "districtcode": 378,
             "statecode": 22,
-            "namelocal": "",
             "dateofbirth": case.get_case_property('date_of_birth') or "",
             "mobileno": case.get_case_property('mobile_number') or "",
-            "mothernamelocal": "",
-            "fathernamelocal": "",
-            "husbandnamelocal": "",
             "email": "",
             "govt_id_name": "",
             "govt_id_number": ""
-        }
+        })
         return json.dumps(data, cls=DjangoJSONEncoder)
 
+    def _setup_names(self, case):
+        data = {}
+        self._setup_name(case.name, 'beneficaryname', 'namelocal', data)
+        self._setup_name(case.get_case_property('mothers_name'), 'mothername', 'mothernamelocal', data)
+        self._setup_name(case.get_case_property('fathers_name'), 'fathername', 'fathernamelocal', data)
+        self._setup_name(case.get_case_property('husbands_name'), 'husbandname', 'husbandnamelocal', data)
+        return data
+
+    def _setup_name(self, name, key_name, key_name_local, data):
+        data[key_name] = ""
+        data[key_name_local] = ""
+        if self._has_special_chars(name):
+            data[key_name_local] = name
+        else:
+            data[key_name] = name
```

Comment:
I know we pass variables by reference and update them inside functions a lot in HQ, but functions are easier to test, and state is easier to track, if you update the dictionary with the function's return value outside the function. e.g. 
```python

    def set_name(name, key_name, key_name_local):
        key = key_name if is_ascii(name) else key_name_local
        return {key: name}

    data = {
        "mothername": "",
        "mothernamelocal: "",
        "fathername": "",
        "fathernamelocal": "",
        "husbandname": "",
        "husbandnamelocal": "",
        # etc.
    }
    for case_property, key_name, key_name_local in [
        ("mothers_name", "mothername", "mothernamelocal"),
        ("fathers_name", "fathername", "fathernamelocal"),
        ("husbands_name", "husbandname", "husbandnamelocal"),
    ]:
        name = case.get_case_property(case_property)
        data.update(set_name(name, key_name, key_name_local))
```

Reply:
> I know we pass variables by reference and update them inside functions a lot in HQ, but functions are easier to test, and state is easier to track

Can I get this on a t-shirt?

Reply:
In this case, `get_payload` is the only public method and others are private. So my understanding is that one would test by just calling `get_payload` and not the individual methods so it's fine here from that perspective?
I liked the for loop, made things much cleaner so used that.